### PR TITLE
feat(tianmu): support select clause without from clause. (#1167, #1168)

### LINKED ISSUE #1167: bug: UNION NULL value returns incorrect result set
### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/stoneatom/stonedb/blob/stonedb-5.7-dev/CONTRIBUTING.md).

### Please confirm if bug report does NOT exists already ?

- [X] I confirm there is no existing issue for this

### Describe the problem

```shell
mysql> create table t2 (a int);
Query OK, 0 rows affected (0.01 sec)

mysql> insert into t2 values (120), (0), (111);
Query OK, 3 rows affected (0.00 sec)
Records: 3  Duplicates: 0  Warnings: 0

mysql> select NULL union select a+0 from t2;
+------+
| NULL |
+------+
| NULL |
+------+
1 row in set (0.00 sec)

```

### Expected behavior

```shell
mysql> select NULL union select a+0 from t2;
+------+
| NULL |
+------+
| NULL |
|  120 |
|    0 |
|  111 |
+------+
4 rows in set (0.00 sec)
```

### How To Reproduce

```sql
create table t2 (a int);
insert into t2 values (120), (0), (111);
select NULL union select a+0 from t2;
```

### Environment

```shell
[root@localhost ~]# /opt/stonedb57/install/bin/mysqld --version
/opt/stonedb57/install/bin/mysqld  Ver 5.7.36-StoneDB for Linux on x86_64 (build-)
build information as follow:
        Repository address: https://github.com/stoneatom/stonedb.git:HEAD
        Branch name: HEAD
        Last commit ID: b44a51ce8
        Last commit time: Date:   Fri Dec 2 16:11:06 2022 +0000
        Build time: Date: Mon Dec  5 06:12:44 UTC 2022
[root@localhost ~]# cat /etc/system-release
CentOS Linux release 7.9.2009 (Core)
```

### Are you interested in submitting a PR to solve the problem?

- [ ] Yes, I will!

### PR DESCRIPTION
**Pull request by lujiashun**
## Summary about this PR

Issue Number: close #1167 #1168

1. add system table sys_tianmu.dummy which has only and must has one row.
2. in sql parse stage, if the select clause without from clause, rewrite the sql, add from clause "from sys_tianmu.dummy"
3. in mysql_test, the system table is initialized in install.db directory,and copy to the mysql_test base_dir, if the file is symbol link, the destination file will be regular file, so change the Path.pm to copy symbol link file;

## Tests Check List

- [ ] Unit test
- [X] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

## Changelog

- [X] New Feature
- [ ] Bug Fix
- [ ] Performance Improvement
- [ ] Build/Testing/CI/CD
- [ ] Documentation
- [ ] Not for changelog (changelog entry is not required)

## Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features

### CODE REVIEW

File: scripts/mysql_system_tables.sql [hunk 1]
```diff
@@ -3372,3 +3372,12 @@ set @had_decomp_dict_table= @@warning_count != 0;
...
+
```
Comment (RingsC) [inline scripts/mysql_system_tables.sql:3375]: ``
DUAL is purely for the convenience of people who require that all [SELECT](https://dev.mysql.com/doc/refman/8.0/en/select.html) statements should have FROM and possibly other clauses. MySQL may ignore the clauses. MySQL does not require FROM DUAL if no tables are referenced.
``

File: scripts/mysql_system_tables.sql [hunk 2]
```diff
@@ -3372,3 +3372,12 @@ set @had_decomp_dict_table= @@warning_count != 0;
...
+DROP TABLE IF EXISTS sys_tianmu.dual;
+CREATE TABLE sys_tianmu.dual (dummy_int int) ENGINE=tianmu;
+
+DROP PROCEDURE IF EXISTS sys_tianmu_wait;
+CREATE DEFINER=root@localhost PROCEDURE sys_tianmu_wait() BEGIN SELECT COUNT(*) INTO @num_rec FROM sys_tianmu.dual;WHILE (@num_rec != 1) DO SELECT COUNT(*) INTO @num_rec FROM sys_tianmu.dual; SELECT SLEEP(1) INTO @sleep;END WHILE;END;
```
Comment (RingsC) [inline scripts/mysql_system_tables.sql]: MTR for this table should be added.

File: sql/sql_parse.cc [hunk 3]
```diff
@@ -5466,6 +5466,57 @@ void mysql_init_multi_delete(LEX *lex)
   lex->query_tables_last= &lex->query_tables;
 }
 
+// Rewrite the query for tianmu, If select_lex has no from clause
+// firstly, chech the tables are all tianmu tables;
```
Comment (RingsC) [inline sql/sql_parse.cc]: It is NOT a good way to do this by adding `dual` to `select_lex`

Comment (lujiashun) [reply sql/sql_parse.cc]: The solution is not graceful , though it seems effective for the issues has been known. 
1. tianmu engines initally does not support select lex without from clause, 
2. We had expanded it but has some limitations and some bugs;for tianmu engine, the temp table requires that it should have tianmu table(pyhisical table) at first, We use the first physical table for this condition, but in some sceanrio , first pyhiscal table is not avaialble to get, for (select null uninao select * from t2);
3. if the select clause is the top level, the current version of tianmu does not support it ,and return to mysql,  it has some problems, first we can not take advantage of the tianmu query engine; secondly, for union clause, it just return the first lex result,result is incorrect ; thirdly, it seems the expanded we have made is just for sp(store procedure);
4. It is a little complex to support select clause without from clause for tianmu, and will be danger to modify the low-level mechanism；
5. this solution is not to modify the low-level mechanism,and to solute it in a high-level, to give every  "select clause without from clause" a fake from cluase, the fake table is similar to dual( in mysql ,the lex remove the dual when parse);
6. Maybe in the future, We can find a better and gracefull solution for this issue, We can put this PR for a while;

Comment (RingsC) [reply sql/sql_parse.cc]: > The solution is not graceful , though it seems effective for the issues has been known.
> 
> 1. tianmu engines initally does not select lex without from clause,
> 2. We had expanded it but has some limitations and some bugs;for tianmu engine, the temp table requires that it should have  tianmu table(pyhisical table) at first, We use the first physical table for this condition, but in some sceanrio , first pyhiscal table is not avaialble to get, for (select null uninao select * from t2);
> 3. if the select clause is the top level, the current version of tianmu does not support it ,and return to mysql,  it has some problems, first we can not take advantage of the tianmu query engine; secondly, for union clause, it just return the first lex result,result is incorrect ; thirdly, it seems the expanded we have made is just for sp(store procedure);
> 4. It is a little complex to support select clause without from clause for tianmu, and will be danger to modify the low-level mechanism；
> 5. this solution is not to modify the low-level mechanism,and to solute it in a high-level, to give every  "select clause without from clause" a fake from cluase, the fake table is similar to dual( in mysql ,the lex remove the dual when parse);
> 6. Maybe in the future, We can find a better and gracefull solution for this issue, We can put this PR for a while;

OK, Let me think it over.

File: storage/tianmu/handler/ha_my_tianmu.cpp [hunk 4]
```diff
@@ -154,6 +154,14 @@ bool ha_my_tianmu_load(THD *thd, sql_exchange *ex, TABLE_LIST *table_list, void
   }
   return false;
 }
+bool ha_my_tianmu_is_tianmu_table(std::string db_name, std::string table_name) {
```
Comment (RingsC) [inline storage/tianmu/handler/ha_my_tianmu.cpp]: we should use table's handler to get which engine it belongs to.  NOT just only by extension name or something else.  AND if the active directory is NOT set to tianmu data or set to somewhere else. the stat will failed.  Overall, it NOT a good choice from my point of view. 
```
  // Notify the storage engines that the tables should be made ready for export.
  for (TABLE_LIST *table_list= all_tables; table_list;
       table_list= table_list->next_global)
  {
    handler *handler_file= table_list->table->file;
    int error= handler_file->extra(HA_EXTRA_EXPORT);
    if (error)
    {
      handler_file->print_error(error, MYF(0));
      return true;
    }
  }
```

Comment (RingsC) [reply storage/tianmu/handler/ha_my_tianmu.cpp]: and how about this case. : (1) using a network file system or OBS, etc... 

stat info: 
https://man7.org/linux/man-pages/man2/lstat.2.html

Comment (lujiashun) [reply storage/tianmu/handler/ha_my_tianmu.cpp]: OK, thanks very much. I will check this and optimize these code.

File: storage/tianmu/handler/ha_my_tianmu.cpp [hunk 5]
```diff
@@ -154,6 +154,14 @@ bool ha_my_tianmu_load(THD *thd, sql_exchange *ex, TABLE_LIST *table_list, void
...
+  std::string path = std::string("./") + db_name + std::string("/") + table_name + common::TIANMU_EXT;
+  struct stat stat_info;
+  int ret = stat(path.c_str(), &stat_info);
+  if (ret != 0)
+    return false;
+  return true;
```
Comment (RingsC) [inline storage/tianmu/handler/ha_my_tianmu.cpp]: BTW: portability is another thing.  `stat` , enmm....

**Comment by lujiashun** [comment]
add sys_tianmu.dummy  as the from clause's table in yacc parse stage.

File: scripts/mysql_system_tables.sql (hunk 1 above)
Comment (lujiashun) [reply scripts/mysql_system_tables.sql:3375]: use dummy table name to avoid of misunderstanding

File: scripts/mysql_system_tables.sql (hunk 2 above)
Comment (lujiashun) [reply scripts/mysql_system_tables.sql]: Fixed. MTR for sys_tianm.dumy is added now.

**Comment by lujiashun** [comment]
for this case, row will be inserted twice. (ref MTR tianmu/func_define.test)
```
DELIMITER //;
CREATE FUNCTION f3(i INT) RETURNS INT DETERMINISTIC
BEGIN
 INSERT INTO t2 VALUES(i+1);
 RETURN 42;
END //
DELIMITER ;//

CREATE TABLE t2(col1 INT);
select f3(121) from sys_tianmu.dummy;
```

```
(gdb) bt
#0  Tianmu::handler::ha_tianmu::write_row (this=0x6751bcc63, buf=0x2d <error: Cannot access memory at address 0x2d>)
    at /data/codebase/stonedb57/stonedb/storage/tianmu/handler/ha_tianmu.cpp:454
#1  0x000055800c66ac7c in handler::ha_write_row (this=0x7f52f417ae80, buf=0x7f5304981340 "\375z") at /data/codebase/stonedb57/stonedb/sql/handler.cc:8189
#2  0x000055800cfb977a in write_record (thd=0x7f52f4000e10, table=0x7f52f4014b10, info=0x7f582a3548a0, update=0x7f582a354920)
    at /data/codebase/stonedb57/stonedb/sql/sql_insert.cc:1904
#3  0x000055800cfb64a8 in Sql_cmd_insert::mysql_insert (this=0x7f52f4204490, thd=0x7f52f4000e10, table_list=0x7f52f4208c58)
    at /data/codebase/stonedb57/stonedb/sql/sql_insert.cc:778
#4  0x000055800cfbd705 in Sql_cmd_insert::execute (this=0x7f52f4204490, thd=0x7f52f4000e10) at /data/codebase/stonedb57/stonedb/sql/sql_insert.cc:3151
#5  0x000055800cd6d948 in mysql_execute_command (thd=0x7f52f4000e10, first_level=false) at /data/codebase/stonedb57/stonedb/sql/sql_parse.cc:3640
#6  0x000055800ccb0fd0 in sp_instr_stmt::exec_core (this=0x7f52f42091e0, thd=0x7f52f4000e10, nextp=0x7f582a355d70)
    at /data/codebase/stonedb57/stonedb/sql/sp_instr.cc:1027
#7  0x000055800ccafbc3 in sp_lex_instr::reset_lex_and_exec_core (this=0x7f52f42091e0, thd=0x7f52f4000e10, nextp=0x7f582a355d70, open_tables=false)
    at /data/codebase/stonedb57/stonedb/sql/sp_instr.cc:452
#8  0x000055800ccb0749 in sp_lex_instr::validate_lex_and_execute_core (this=0x7f52f42091e0, thd=0x7f52f4000e10, nextp=0x7f582a355d70, open_tables=false)
    at /data/codebase/stonedb57/stonedb/sql/sp_instr.cc:753
\#9  0x000055800ccb0cc4 in sp_instr_stmt::execute (this=0x7f52f42091e0, thd=0x7f52f4000e10, nextp=0x7f582a355d70)
    at /data/codebase/stonedb57/stonedb/sql/sp_instr.cc:938
#10 0x000055800cca8de1 in sp_head::execute (this=0x7f52f4202590, thd=0x7f52f4000e10, merge_da_on_success=true)
    at /data/codebase/stonedb57/stonedb/sql/sp_head.cc:796
#11 0x000055800ccaa2da in sp_head::execute_function (this=0x7f52f4202590, thd=0x7f52f4000e10, argp=0x7f52f4005a08, argcount=1, return_value_fld=0x7f52f417c820)
    at /data/codebase/stonedb57/stonedb/sql/sp_head.cc:1287
#12 0x000055800c70e235 in Item_func_sp::execute_impl (this=0x7f52f4005960, thd=0x7f52f4000e10) at /data/codebase/stonedb57/stonedb/sql/item_func.cc:8706
#13 0x000055800c70df7d in Item_func_sp::execute (this=0x7f52f4005960) at /data/codebase/stonedb57/stonedb/sql/item_func.cc:8637
#14 0x000055800c712676 in Item_func_sp::val_int (this=0x7f52f4005960) at /data/codebase/stonedb57/stonedb/sql/item_func.h:2741
#15 0x000055800daa8baf in Tianmu::core::MysqlExpression::ItemInt2ValueOrNull (item=0x7f52f4005960)
    at /data/codebase/stonedb57/stonedb/storage/tianmu/core/mysql_expression.cpp:561
#16 0x000055800daa82b4 in Tianmu::core::MysqlExpression::Evaluate (this=0x7f52f40158c0)
    at /data/codebase/stonedb57/stonedb/storage/tianmu/core/mysql_expression.cpp:459
#17 0x000055800d875eba in Tianmu::vcolumn::ExpressionColumn::IsNullImpl (this=0x7f52f40113d0, mit=...)
    at /data/codebase/stonedb57/stonedb/storage/tianmu/vc/expr_column.cpp:127
#18 0x000055800d758ea0 in Tianmu::vcolumn::VirtualColumnBase::IsNull (this=0x7f52f40113d0, mit=...)
    at /data/codebase/stonedb57/stonedb/storage/tianmu/vc/virtual_column_base.h:112
#19 0x000055800d76e20c in Tianmu::core::TempTable::SendResult (this=0x7f52f417ed10, limit=1, offset=0, sender=..., pagewise=false)
    at /data/codebase/stonedb57/stonedb/storage/tianmu/core/temp_table_low.cpp:474
#20 0x000055800d76d26e in Tianmu::core::TempTable::FillMaterializedBuffers (this=0x7f52f417ed10, local_limit=1, local_offset=0, sender=0x7f52f420fef0,
    pagewise=false) at /data/codebase/stonedb57/stonedb/storage/tianmu/core/temp_table_low.cpp:317
#21 0x000055800d75527f in Tianmu::core::TempTable::Materialize (this=0x7f52f417ed10, in_subq=false, sender=0x7f52f420fef0, lazy=false)
    at /data/codebase/stonedb57/stonedb/storage/tianmu/core/temp_table.cpp:2064
#22 0x000055800d6d6dfe in Tianmu::core::Engine::Execute (this=0x558010611800, thd=0x7f52f4000e10, lex=0x7f52f4003138, result_output=0x7f52f4006ad8,
    unit_for_union=0x0) at /data/codebase/stonedb57/stonedb/storage/tianmu/core/engine_execute.cpp:482
#23 0x000055800d6d5a31 in Tianmu::core::Engine::HandleSelect (this=0x558010611800, thd=0x7f52f4000e10, lex=0x7f52f4003138,
    result=@0x7f582a358488: 0x7f52f4006ad8, setup_tables_done_option=0, res=@0x7f582a35847c: 32594, optimize_after_tianmu=@0x7f582a358484: 1,
    tianmu_free_join=@0x7f582a358480: 1, with_insert=0) at /data/codebase/stonedb57/stonedb/storage/tianmu/core/engine_execute.cpp:238
#24 0x000055800d7ef87c in Tianmu::handler::ha_my_tianmu_query (thd=0x7f52f4000e10, lex=0x7f52f4003138, result_output=@0x7f582a358488: 0x7f52f4006ad8,
    setup_tables_done_option=0, res=@0x7f582a35847c: 32594, optimize_after_tianmu=@0x7f582a358484: 1, tianmu_free_join=@0x7f582a358480: 1, with_insert=0)
    at /data/codebase/stonedb57/stonedb/storage/tianmu/handler/ha_my_tianmu.cpp:94
--Type <RET> for more, q to quit, c to continue without paging--\
#25 0x000055800cd728d4 in execute_sqlcom_select (thd=0x7f52f4000e10, all_tables=0x7f52f417bc10) at /data/codebase/stonedb57/stonedb/sql/sql_parse.cc:5208
#26 0x000055800cd6b943 in mysql_execute_command (thd=0x7f52f4000e10, first_level=true) at /data/codebase/stonedb57/stonedb/sql/sql_parse.cc:2851
#27 0x000055800cd73962 in mysql_parse (thd=0x7f52f4000e10, parser_state=0x7f582a359630) at /data/codebase/stonedb57/stonedb/sql/sql_parse.cc:5646
#28 0x000055800cd684f3 in dispatch_command (thd=0x7f52f4000e10, com_data=0x7f582a359ee0, command=COM_QUERY)
    at /data/codebase/stonedb57/stonedb/sql/sql_parse.cc:1495
#29 0x000055800cd6736a in do_command (thd=0x7f52f4000e10) at /data/codebase/stonedb57/stonedb/sql/sql_parse.cc:1034
#30 0x000055800ceaff2c in handle_connection (arg=0x558015275d20) at /data/codebase/stonedb57/stonedb/sql/conn_handler/connection_handler_per_thread.cc:313
#31 0x000055800d5fc0a3 in pfs_spawn_thread (arg=0x55801502e8e0) at /data/codebase/stonedb57/stonedb/storage/perfschema/pfs.cc:2197
#32 0x00007f583187d609 in start_thread (arg=<optimized out>) at pthread_create.c:477
#33 0x00007f5831651133 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

```
(gdb) bt
#0  Tianmu::handler::ha_tianmu::write_row (this=0x66bc4cf8e, buf=0x2d <error: Cannot access memory at address 0x2d>)
    at /data/codebase/stonedb57/stonedb/storage/tianmu/handler/ha_tianmu.cpp:454
#1  0x000055800c66ac7c in handler::ha_write_row (this=0x7f52f417ae80, buf=0x7f5304981340 "\375z") at /data/codebase/stonedb57/stonedb/sql/handler.cc:8189
#2  0x000055800cfb977a in write_record (thd=0x7f52f4000e10, table=0x7f52f4014b10, info=0x7f582a3548a0, update=0x7f582a354920)
    at /data/codebase/stonedb57/stonedb/sql/sql_insert.cc:1904
#3  0x000055800cfb64a8 in Sql_cmd_insert::mysql_insert (this=0x7f52f4204490, thd=0x7f52f4000e10, table_list=0x7f52f4208c58)
    at /data/codebase/stonedb57/stonedb/sql/sql_insert.cc:778
#4  0x000055800cfbd705 in Sql_cmd_insert::execute (this=0x7f52f4204490, thd=0x7f52f4000e10) at /data/codebase/stonedb57/stonedb/sql/sql_insert.cc:3151
#5  0x000055800cd6d948 in mysql_execute_command (thd=0x7f52f4000e10, first_level=false) at /data/codebase/stonedb57/stonedb/sql/sql_parse.cc:3640
#6  0x000055800ccb0fd0 in sp_instr_stmt::exec_core (this=0x7f52f42091e0, thd=0x7f52f4000e10, nextp=0x7f582a355d70)
    at /data/codebase/stonedb57/stonedb/sql/sp_instr.cc:1027
#7  0x000055800ccafbc3 in sp_lex_instr::reset_lex_and_exec_core (this=0x7f52f42091e0, thd=0x7f52f4000e10, nextp=0x7f582a355d70, open_tables=false)
    at /data/codebase/stonedb57/stonedb/sql/sp_instr.cc:452
#8  0x000055800ccb0749 in sp_lex_instr::validate_lex_and_execute_core (this=0x7f52f42091e0, thd=0x7f52f4000e10, nextp=0x7f582a355d70, open_tables=false)
    at /data/codebase/stonedb57/stonedb/sql/sp_instr.cc:753
#9  0x000055800ccb0cc4 in sp_instr_stmt::execute (this=0x7f52f42091e0, thd=0x7f52f4000e10, nextp=0x7f582a355d70)
    at /data/codebase/stonedb57/stonedb/sql/sp_instr.cc:938
#10 0x000055800cca8de1 in sp_head::execute (this=0x7f52f4202590, thd=0x7f52f4000e10, merge_da_on_success=true)
    at /data/codebase/stonedb57/stonedb/sql/sp_head.cc:796
#11 0x000055800ccaa2da in sp_head::execute_function (this=0x7f52f4202590, thd=0x7f52f4000e10, argp=0x7f52f4005a08, argcount=1, return_value_fld=0x7f52f417c820)
    at /data/codebase/stonedb57/stonedb/sql/sp_head.cc:1287
#12 0x000055800c70e235 in Item_func_sp::execute_impl (this=0x7f52f4005960, thd=0x7f52f4000e10) at /data/codebase/stonedb57/stonedb/sql/item_func.cc:8706
#13 0x000055800c70df7d in Item_func_sp::execute (this=0x7f52f4005960) at /data/codebase/stonedb57/stonedb/sql/item_func.cc:8637
#14 0x000055800c712676 in Item_func_sp::val_int (this=0x7f52f4005960) at /data/codebase/stonedb57/stonedb/sql/item_func.h:2741
#15 0x000055800daa8baf in Tianmu::core::MysqlExpression::ItemInt2ValueOrNull (item=0x7f52f4005960)
    at /data/codebase/stonedb57/stonedb/storage/tianmu/core/mysql_expression.cpp:561
#16 0x000055800daa82b4 in Tianmu::core::MysqlExpression::Evaluate (this=0x7f52f40158c0)
    at /data/codebase/stonedb57/stonedb/storage/tianmu/core/mysql_expression.cpp:459
#17 0x000055800d875dd2 in Tianmu::vcolumn::ExpressionColumn::GetValueInt64Impl (this=0x7f52f40113d0, mit=...)
    at /data/codebase/stonedb57/stonedb/storage/tianmu/vc/expr_column.cpp:119
#18 0x000055800d758e6a in Tianmu::vcolumn::VirtualColumnBase::GetValueInt64 (this=0x7f52f40113d0, mit=...)
    at /data/codebase/stonedb57/stonedb/storage/tianmu/vc/virtual_column_base.h:93
#19 0x000055800d76e260 in Tianmu::core::TempTable::SendResult (this=0x7f52f417ed10, limit=1, offset=0, sender=..., pagewise=false)
    at /data/codebase/stonedb57/stonedb/storage/tianmu/core/temp_table_low.cpp:477
#20 0x000055800d76d26e in Tianmu::core::TempTable::FillMaterializedBuffers (this=0x7f52f417ed10, local_limit=1, local_offset=0, sender=0x7f52f420fef0,
    pagewise=false) at /data/codebase/stonedb57/stonedb/storage/tianmu/core/temp_table_low.cpp:317
#21 0x000055800d75527f in Tianmu::core::TempTable::Materialize (this=0x7f52f417ed10, in_subq=false, sender=0x7f52f420fef0, lazy=false)
    at /data/codebase/stonedb57/stonedb/storage/tianmu/core/temp_table.cpp:2064
#22 0x000055800d6d6dfe in Tianmu::core::Engine::Execute (this=0x558010611800, thd=0x7f52f4000e10, lex=0x7f52f4003138, result_output=0x7f52f4006ad8,
    unit_for_union=0x0) at /data/codebase/stonedb57/stonedb/storage/tianmu/core/engine_execute.cpp:482
#23 0x000055800d6d5a31 in Tianmu::core::Engine::HandleSelect (this=0x558010611800, thd=0x7f52f4000e10, lex=0x7f52f4003138,
    result=@0x7f582a358488: 0x7f52f4006ad8, setup_tables_done_option=0, res=@0x7f582a35847c: 32594, optimize_after_tianmu=@0x7f582a358484: 1,
    tianmu_free_join=@0x7f582a358480: 1, with_insert=0) at /data/codebase/stonedb57/stonedb/storage/tianmu/core/engine_execute.cpp:238
#24 0x000055800d7ef87c in Tianmu::handler::ha_my_tianmu_query (thd=0x7f52f4000e10, lex=0x7f52f4003138, result_output=@0x7f582a358488: 0x7f52f4006ad8,
    setup_tables_done_option=0, res=@0x7f582a35847c: 32594, optimize_after_tianmu=@0x7f582a358484: 1, tianmu_free_join=@0x7f582a358480: 1, with_insert=0)
    at /data/codebase/stonedb57/stonedb/storage/tianmu/handler/ha_my_tianmu.cpp:94
--Type <RET> for more, q to quit, c to continue without paging--
#25 0x000055800cd728d4 in execute_sqlcom_select (thd=0x7f52f4000e10, all_tables=0x7f52f417bc10) at /data/codebase/stonedb57/stonedb/sql/sql_parse.cc:5208
#26 0x000055800cd6b943 in mysql_execute_command (thd=0x7f52f4000e10, first_level=true) at /data/codebase/stonedb57/stonedb/sql/sql_parse.cc:2851
#27 0x000055800cd73962 in mysql_parse (thd=0x7f52f4000e10, parser_state=0x7f582a359630) at /data/codebase/stonedb57/stonedb/sql/sql_parse.cc:5646
#28 0x000055800cd684f3 in dispatch_command (thd=0x7f52f4000e10, com_data=0x7f582a359ee0, command=COM_QUERY)
    at /data/codebase/stonedb57/stonedb/sql/sql_parse.cc:1495
#29 0x000055800cd6736a in do_command (thd=0x7f52f4000e10) at /data/codebase/stonedb57/stonedb/sql/sql_parse.cc:1034
#30 0x000055800ceaff2c in handle_connection (arg=0x558015275d20) at /data/codebase/stonedb57/stonedb/sql/conn_handler/connection_handler_per_thread.cc:313
#31 0x000055800d5fc0a3 in pfs_spawn_thread (arg=0x55801502e8e0) at /data/codebase/stonedb57/stonedb/storage/perfschema/pfs.cc:2197
#32 0x00007f583187d609 in start_thread (arg=<optimized out>) at pthread_create.c:477
#33 0x00007f5831651133 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95

```

File: storage/tianmu/vc/expr_column.cpp [hunk 6]
```diff
@@ -122,6 +122,12 @@ int64_t ExpressionColumn::GetValueInt64Impl(const core::MIIterator &mit) {
   return last_val_->Get64();
 }
 
+int64_t ExpressionColumn::GetValueInt64Direct(const core::MIIterator &mit) {
```
Comment (wisehead) [inline storage/tianmu/vc/expr_column.cpp:125]: can we use ExpressionColumn::GetValueInt64Impl() to get the value from dummy table?

**Review by wisehead — CHANGES_REQUESTED** [review]

File: storage/tianmu/vc/expr_column.cpp (hunk 6 above)
Comment (lujiashun) [reply storage/tianmu/vc/expr_column.cpp:125]: please refer to: https://github.com/stoneatom/stonedb/pull/1247#issuecomment-1427520408
the function is not used to get dummy table value;  and this modification is  for SP( user defined function）thas has dml sql.

check function TempTable::SendResult(int64_t limit, int64_t offset, ResultSender &sender, bool pagewise) as below，
SP is evaluated once （insert one row）in LINE474, SP is evaluated the second time（insert one row again) in LINE477 (GetValueInt64), 

GetValueInt64Direct means just get value without evaluating;

The twice insertted row is trigged by this PR, it has the issue even without the PR;
 ```
LINE470:  if (ct == common::ColumnType::INT || ct == common::ColumnType::MEDIUMINT || ct == common::ColumnType::SMALLINT ||
LINE471:      ct == common::ColumnType::BYTEINT || ct == common::ColumnType::NUM || ct == common::ColumnType::BIGINT ||
LINE472:      ct == common::ColumnType::BIT) {
LINE473:    auto data_ptr = new types::TianmuNum();
LINE474:    if (vc->IsNull(it))
LINE475:      data_ptr->SetToNull();
LINE476:    else
LINE477:      data_ptr->Assign(vc->GetValueInt64(it), col->Type().GetScale(), false, ct);
LINE478:    record.emplace_back(data_ptr);
```

Comment (wisehead) [reply storage/tianmu/vc/expr_column.cpp:125]: please add some comments for these new added functions, to expain why they are added.

Comment (lujiashun) [reply storage/tianmu/vc/expr_column.cpp:125]: OK,thanks. Comments are added in the file virtual_column_base.h，please check it.

**Review by wisehead — APPROVED** [review]

**Pull request closed without merging**